### PR TITLE
Support uniqueItems constraint

### PR DESCRIPTION
**Pull request by ianhoffman**
Supports uniqueConstraints on arrays of integers and strings, closing out #42.

Note that I didn't run Hackfmt because it was creating all sorts of diffs in unrelated files. Not sure what's going on here — I'm on HHVM 4.56.4, if that's any help.

## Implementation

My approach is to update `ArrayBuilder` such that the generated `check()` method calls into a new `ArrayUniqueItemsConstraint` validator when appropriate. The `ArrayUniqueItemsConstraint` validator is responsible for validating that all items in the input are unique and coercing them to a keyset.

I opted not to set a `uniqueItems` property on the validator generated by `ArrayBuilder` because this property wouldn't be used for anything. Instead, we only bother generating the call to `ArrayUniqueItemsConstraint` if `uniqueItems` is set to true in the schema. The only use-case I can see for generating a `uniqueItems` property is to support introspection in runtime code.

Finally, we need to validate that `uniqueItems` is supported for the given array and item type. For now, this means validating that the given item type is an arraykey, since these are the only types which we can stick in a keyset in Hack. However, to get the item type, we need to read from the SchemaBuilder for the item. This builder, which is called the `singleItemSchemaBuilder` in the `ArrayBuilder` code, is instantiated in middle of the ArrayBuilder's `build` method. I found this frustrating since I wanted to be able to reject unsupported `uniqueItems` constraints as the first thing we do when calling `build()`. As such, I decided to refactor so that the first two things we do when calling `build()` are set up the `singleItemSchemaBuilder` and then use it to validate the `uniqueItems` constraint. This means that for the rest of the call to `build()` we can assume that `uniqueItems` is valid and don't need to bother checking that the item type is an arraykey.

## Testing

I added several unit tests and also manually tested that we reject unsupported uniqueItems constraints (e.g., when the item type is a boolean).

### CODE REVIEW

File: src/Constraints/ArrayUniqueItemsConstraint.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+use namespace HH\Lib\C;
+use namespace Slack\Hack\JsonSchema;
+
+class ArrayUniqueItemsConstraint {
+  public static function check<T as arraykey>(
+    vec<T> $input,
+    string $pointer,
+  ): keyset<T> {
+    $output = keyset($input);
```
Comment (mwildehahn) [inline src/Constraints/ArrayUniqueItemsConstraint.php:13]: One thought is that we could check for `coerce` here and if so just return `keyset($input)`. I can't think of a lot of cases where we want to ensure that unique items are provided and reject it vs accepting an array but having keyset called already on it.

**Review by mwildehahn — APPROVED** [review]
This looks great! Good call on not storing the `unique_items` prop since we don't need that 🙌 . 

The hackfmt issues are likely because this hasn't been run under a newer version of hhvm. Let me run that with 4.56 and commit that so we can run it for this without altering the change set.

File: src/Constraints/ArrayUniqueItemsConstraint.php (hunk 1 above)
Comment (ianhoffman) [reply src/Constraints/ArrayUniqueItemsConstraint.php:13]: Hmm, I guess my only concern there is we're overloading `coerce` from its usual meaning. But I agree that deduping items is a more common use-case than rejecting input which contains duplicates, so it'd be good to support somehow, and I can't think of a better way currently. So yeah, I'm inclined to make this change.

Comment (ianhoffman) [reply src/Constraints/ArrayUniqueItemsConstraint.php:14]: @mwildehahn updated in https://github.com/slackhq/hack-json-schema/pull/43/commits/edf693cf16ac75adfe52eb86cbc2afd4d27519df

File: tests/examples/empty-schema.json [hunk 2]
```diff
@@ -1 +1,2 @@
-{}
```
Comment (ianhoffman) [inline tests/examples/empty-schema.json:1]: This got updated by hackfmt... 🤷‍♂️ ?

Comment (ianhoffman) [reply tests/examples/empty-schema.json:1]: I think we should update `make format` to skip non-php/Hack files.

**Pull request merged**